py/bitbox02/bitbox02/communication/bitbox_api_protocol.py

```diff
@@ -179,46 +180,69 @@ def set_app_static_privkey(self, privkey: bytes) -> None:
+        super().__init__()
         self._transport = transport
+        self._noise: NoiseConnection = None
 
-        self.version = parse_device_version(serial_number)
-        if self.version is None:
-            self.close()
-            raise ValueError(f"Could not parse version from {serial_number}")
-        # Delete the prelease part, as it messes with the comparison (e.g. 3.0.0-pre < 3.0.0 is
-        # True, but the 3.0.0-pre has already the same API breaking changes like 3.0.0...).
-        self.version = semver.VersionInfo(
-            self.version.major, self.version.minor, self.version.patch, build=self.version.build
-        )
+    def close(self) -> None:
+        self._transport.close()
 
-        if self.version >= semver.VersionInfo(2, 0, 0):
-            noise_config.attestation_check(self._perform_attestation())
+    def _raw_query(self, msg: bytes) -> bytes:
+        cid = self._transport.generate_cid()
+        return self._transport.query(msg, HWW_CMD, cid)
 
-            # Invoke unlock workflow on the device.
-            # In version <2.0.0, the device did this automatically.
-            unlock_result = self._query(OP_UNLOCK)
-            if self.version < semver.VersionInfo(3, 0, 0):
-                assert unlock_result == b""
-            else:
-                # since 3.0.0, unlock can fail if cancelled
-                if unlock_result == RESPONSE_FAILURE:
-                    self.close()
-                    raise Exception("Unlock process aborted")
+    def query(self, cmd: bytes, msg_data: bytes) -> Tuple[bytes, bytes]:
+        """
+        Encapsulates the given OP_* command and message data into a packet,
+        and unpacks the response status code and data.
+        """
+        response = self._raw_query(cmd + msg_data)
+        return response[0:1], response[1:]
 
-        self.noise = self._create_noise_channel(noise_config)
+    @abstractmethod
+    def _encode_noise_request(self, encrypted_msg: bytes) -> bytes:
+        """ Encapsulates an OP_NOISE_MSG message. """
+        ...
+
+    def _decode_noise_response(self, encrypted_msg: bytes) -> Tuple[bytes, bytes]:
+        """ De-encapsulate an OP_NOISE_MSG response. """
+        ...
+
+    @abstractmethod
+    def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
+        """
+        Executes a OP_HER_COMEZ_TEH_HANDSHAEK query with the given
+        request data.
+        Returns a pair (response status, response data).
+        """
+        ...
+
+    def encrypted_query(self, msg: bytes) -> bytes:
+        """
+        Sends msg bytes and reads response bytes over an encrypted channel.
+        """
+        encrypted_msg = self._noise.encrypt(msg)
+        encrypted_msg = self._encode_noise_request(encrypted_msg)
+
+        response = self._raw_query(encrypted_msg)
+        response_status, response = self._decode_noise_response(response)
+        if response_status != RESPONSE_SUCCESS:
```

Comment:
same as below, if this is moved into `_decode_noise_response`, the function does not need to return a fake response in V1-V6.

Reply:
But it can return failure (which we detect if we receive back an empty response) - this avoids throwing the same exception from two points.